cpp/src/io/parquet/chunk_dict.cu

```diff
@@ -102,82 +112,74 @@ __global__ void __launch_bounds__(block_size, 1)
   auto chunk   = frag.chunk;
   auto col     = chunk->col_desc;
 
-  size_type start_row = frag.start_row;
-  size_type end_row   = frag.start_row + frag.num_rows;
+  if (not chunk->use_dictionary) { return; }
 
-  __shared__ size_type s_start_value_idx;
-  __shared__ size_type s_num_values;
+  __shared__ cg::experimental::block_tile_memory<4, block_size> shared;
+  auto thb   = cg::experimental::this_thread_block(shared);
+  auto block = cg::experimental::tiled_partition<block_size>(thb);
+  // cub::BlockReduce instead of cg::reduce for performance purpose
+  using block_reduce = cub::BlockReduce<size_type, block_size>;
+  __shared__ typename block_reduce::TempStorage reduce_storage;
 
-  if (not chunk->use_dictionary) { return; }
+  size_type start_row = frag.start_row;
+  size_type end_row   = frag.start_row + frag.num_rows;
 
-  if (t == 0) {
-    // Find the bounds of values in leaf column to be inserted into the map for current chunk
-    auto cudf_col      = *(col->parent_column);
-    s_start_value_idx  = row_to_value_idx(start_row, cudf_col);
-    auto end_value_idx = row_to_value_idx(end_row, cudf_col);
-    s_num_values       = end_value_idx - s_start_value_idx;
-  }
-  __syncthreads();
+  // Find the bounds of values in leaf column to be inserted into the map for current chunk
+  auto const cudf_col          = *(col->parent_column);
+  auto const s_start_value_idx = row_to_value_idx(start_row, cudf_col);
+  auto const end_value_idx     = row_to_value_idx(end_row, cudf_col);
 
   column_device_view const& data_col = *col->leaf_column;
-  using block_reduce                 = cub::BlockReduce<size_type, block_size>;
-  __shared__ typename block_reduce::TempStorage reduce_storage;
 
   // Make a view of the hash map
   auto hash_map_mutable = map_type::device_mutable_view(
     chunk->dict_map_slots, chunk->dict_map_size, KEY_SENTINEL, VALUE_SENTINEL);
-  auto hash_map = map_type::device_view(
-    chunk->dict_map_slots, chunk->dict_map_size, KEY_SENTINEL, VALUE_SENTINEL);
 
   __shared__ int total_num_dict_entries;
-  for (size_type i = 0; i < s_num_values; i += block_size) {
-    // add the value to hash map
-    size_type val_idx = i + t + s_start_value_idx;
-    bool is_valid =
-      (i + t < s_num_values && val_idx < data_col.size()) and data_col.is_valid(val_idx);
+  auto val_idx = s_start_value_idx + t;
+  while (block.any(val_idx < end_value_idx)) {
```

Comment:
I think this does the same thing without requiring communication. 
```suggestion
  while( val_idx - block.size() < end_value_idx) {
```

Reply:
wow 
That's really neat.

Reply:
Personally, thought the `while (block.any(val_idx < end_value_idx)) {` is more readable than the for loop, the for loop was more redable than `while (block.any(val_idx < end_value_idx)) {`

Reply:
> Personally, thought the `while (block.any(val_idx < end_value_idx)) {` is more readable than the for loop, the for loop was more redable than `while (block.any(val_idx < end_value_idx)) {`

I'm a little confused by what you mean since it sounds like you're contradicting yourself ;) 

`while (block.any(val_idx < end_value_idx))` is definitely more readable, but the `any` call definitely isn't free as it requires intra and inter-warp communication.

Reply:
`cg::any` Removed.

Reply:
Typo. I meant that the order of readability to me are (best to worst)
1. While (any 
2. The for loop that existed
3. The suggested posted here

Reply:
Updates: adopt @devavret 's idea and revert back to `while(block.any(...))` loop. By doing so, we also get rid of the block-wide sync at the end of the loop and the code is more readable.